go/consensus/pricediscovery/pricediscovery.go

```diff
@@ -0,0 +1,161 @@
+// Package pricediscovery implements dynamic price discovery.
+package pricediscovery
+
+import (
+	"context"
+	"sync"
+
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/common/quantity"
+	consensus "github.com/oasisprotocol/oasis-core/go/consensus/api"
+)
+
+const (
+	// defaultWindowSize is the default number of recent blocks to use for calculating min gas
+	// price.
+	//
+	// NOTE: Code assumes that this is relatively small.
+	defaultWindowSize uint64 = 6
+
+	// defaultFullBlockThreshold is the default percentage of block used gas over which a block
+	// should be considered full.
+	defaultFullBlockThreshold float64 = 0.5
+)
+
+type priceDiscovery struct {
+	mu sync.RWMutex
+
+	fallbackGasPrice *quantity.Quantity
+	minGasPrice      *quantity.Quantity
+	computedGasPrice *quantity.Quantity
+
+	// blockPrices is a rolling-array containing minimum transaction prices for last up to
+	// `windowSize` blocks.
+	blockPrices []*quantity.Quantity
+	// tracks the current index of the blockPrices rolling array.
+	blockPricesCurrentIdx int
+
+	windowSize         uint64
+	fullBlockThreshold float64
+
+	client consensus.ClientBackend
+
+	logger *logging.Logger
+}
+
+// GasPrice implements consensus.PriceDiscovery.
+func (pd *priceDiscovery) GasPrice(_ context.Context) (*quantity.Quantity, error) {
+	pd.mu.RLock()
+	defer pd.mu.RUnlock()
+
+	// Choose the maximum of (fallback, min, computed) gas prices.
```

Comment:
This could also be computed after every block. Here, we would just serve the computed value.